### PR TITLE
Added JsCreatePromise to JsRT

### PR DESCRIPTION
**Pull request by kfarnung**

### CODE REVIEW

**Comment by liminzhu** [comment]
+@curtisman , this is the context https://github.com/Microsoft/ChakraCore/issues/2541

File: lib/Jsrt/ChakraCommon.h [hunk 1]
```diff
@@ -2378,6 +2378,30 @@ typedef unsigned short uint16_t;
...
 
+    /// <summary>
+    ///     Creates a new JavaScript Promise object.
+    /// </summary>
+    /// <remarks>
+    ///     <para>
+    ///     Requires an active script context.
+    ///     </para>
+    /// </remarks>
+    /// <param name="promise">The new Promise object.</param>
```
Comment (boingoing) [inline lib/Jsrt/ChakraCommon.h]: Just a nitpick, this param tag is all one line while the others are split up.

File: lib/Jsrt/Jsrt.cpp [hunk 2]
```diff
@@ -2866,6 +2867,32 @@ CHAKRA_API JsSetPromiseContinuationCallback(_In_ JsPromiseContinuationCallback p
...
+    Js::JavascriptPromise *jsPromise = scriptContext->GetLibrary()->CreatePromise();
+    Js::JavascriptPromise::InitializePromise(jsPromise, &jsResolve, &jsReject, scriptContext);
+
+    *promise = (JsValueRef)jsPromise;
+    *resolve = (JsValueRef)jsResolve;
+    *reject = (JsValueRef)jsReject;
+
+    return JsNoError;
+  });
+}
```
Comment (dilijev) [inline lib/Jsrt/Jsrt.cpp]: nit: indentation level in the rest of this file is 4 spaces.

Comment (kfarnung) [inline lib/Jsrt/Jsrt.cpp]: Good catch, VS doesn't seem to auto-detect indentation.

File: lib/Jsrt/Jsrt.cpp [hunk 3]
```diff
@@ -2866,6 +2867,32 @@ CHAKRA_API JsSetPromiseContinuationCallback(_In_ JsPromiseContinuationCallback p
     /*allowInObjectBeforeCollectCallback*/true);
 }
 
+CHAKRA_API JsCreatePromise(_Out_ JsValueRef *promise, _Out_ JsValueRef *resolve, _Out_ JsValueRef *reject)
+{
+  return ContextAPIWrapper<true>([&](Js::ScriptContext *scriptContext, TTDRecorder& _actionEntryPopper) -> JsErrorCode {
```
Comment (boingoing) [inline lib/Jsrt/Jsrt.cpp]: Nitpick, indentation here.

File: lib/Jsrt/ChakraCommon.h (hunk 1 above)
Comment (dilijev) [inline lib/Jsrt/ChakraCommon.h]: We're inconsistent about this elsewhere in this file as well. I wonder if it has to do with line length and VS doing something automatically with formatting?

**Comment by liminzhu** [comment]
I'm a little confused about what `resolveFunction` and `rejectFunction` output are and how I would use them.

Let's say JsCreatePromise is equivalent to `new Promise((res, rej) => {...})`, would this make more sense? 
```
JsCreatePromise(_In_ JsValueRef executor, _Out_ JsValueRef *promise)
``` 
`executor` could be a native function created by `JsCreateFunction`.

**Comment by boingoing** [comment]
@liminzhu 

> I'm a little confused about what resolveFunction and rejectFunction output are and how I would use them.

These are essentially the same functions passed to the executor. Another method to fetch them might be like this:

```JavaScript
var resolve, reject;
function executor(res, rej) { resolve = res; reject = rej; };
var promise = new Promise(executor);
```
Goal would be to allow our native code to resolve the promise whenever it wants to. One use might be an API which reads a file from the network and returns a promise resolved with the contents of that file. Native code can construct the promise using JsCreatePromise and hold-on to resolveFunction for that promise. Assume the native-side performs some processing to fetch the result data and when that is done it can call the resolveFunction with that result data which will handle resolving the promise in the engine. I guess the key thing here is that the native code can resolve the promise in an async way without having to create an executor function which would manually suck out the resolve/reject functions.

**Comment by liminzhu** [comment]
Makes much sense now. Thanks @boingoing !

File: lib/Jsrt/Jsrt.cpp [hunk 4]
```diff
@@ -4508,4 +4509,30 @@ CHAKRA_API JsRunSerialized(
...
+CHAKRA_API JsCreatePromise(_Out_ JsValueRef *promise, _Out_ JsValueRef *resolve, _Out_ JsValueRef *reject)
+{
+    return ContextAPIWrapper<true>([&](Js::ScriptContext *scriptContext, TTDRecorder& _actionEntryPopper) -> JsErrorCode {
+        PERFORM_JSRT_TTD_RECORD_ACTION_NOT_IMPLEMENTED(scriptContext);
+
+        PARAM_NOT_NULL(promise);
+        PARAM_NOT_NULL(resolve);
+        PARAM_NOT_NULL(reject);
+
+        *promise = nullptr;
```
Comment (obastemur) [inline lib/Jsrt/Jsrt.cpp]: is this needed? (same applies to two below)

Comment (galvesribeiro) [inline lib/Jsrt/Jsrt.cpp]: It is not mandatory to have either resolve() or reject() methods FYI.

**Comment by galvesribeiro** [comment]
LGTM. I just would like to mention that it is not required to have both reject() and response() methods. And if possible, it would be good to accept multiple methods accepted because people can catch multiple types of errors on diff methods. For example:

`mypromise.then(...).then(...).catch(...).catch(...)` <- that is totally viable.

**Comment by kfarnung** [comment]
@galvesribeiro I think you might have it backwards, the `resolve` and `reject` parameters are out parameters that will contain pointers to the functions that the native code will call when it has either resolved or rejected the promise.  These are equivalent to the parameters that come into the `executor` method when creating a promise in JavaScript.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise

For continuations you would just use the JsRT functions to get the property (`then`, `catch`, etc.) and then call it normally with `JsCallFunction`.

In C#:
```
JavaScriptPropertyId thenId = JavaScriptPropertyId.FromString("then");
JavaScriptValue thenFunction = promise.GetProperty(thenId);

JavaScriptValue resolveFunc = JavaScriptValue.CreateFunction(ResolveCallback);
JavaScriptValue rejectFunc = JavaScriptValue.CreateFunction(RejectCallback);
JavaScriptValue thenPromise = thenFunction.CallFunction(promise, resolveFunc, rejectFunc);
```

Since it's just a JavaScript function, you can choose to leave out the second parameter when calling it.

File: lib/Jsrt/Jsrt.cpp (hunk 4 above)
Comment (kfarnung) [inline lib/Jsrt/Jsrt.cpp]: @obastemur I was just following the pattern of other methods.  The idea was to ensure that regardless of how the flow changed we would always have set some value.  I realize that in this flow we will always set a value, so I'm fine removing it if others agree.

**Comment by galvesribeiro** [comment]
Oh! Right, I understood it completely wrong, sorry. You are right! Good work!

File: lib/Jsrt/Jsrt.cpp (hunk 4 above)
Comment (dilijev) [inline lib/Jsrt/Jsrt.cpp]: I don't see a reason to remove this. In this case the compiler will be able to figure out that this operation is not needed -- if the flow ever changes, we are protected.

File: lib/Jsrt/Jsrt.cpp [hunk 5]
```diff
@@ -4523,4 +4524,30 @@ CHAKRA_API JsRunSerialized(
         sourceContext, // use the same user provided sourceContext as scriptLoadSourceContext
         buffer, bufferVal, sourceContext, url, false, result);
 }
+
+CHAKRA_API JsCreatePromise(_Out_ JsValueRef *promise, _Out_ JsValueRef *resolve, _Out_ JsValueRef *reject)
+{
+    return ContextAPIWrapper<true>([&](Js::ScriptContext *scriptContext, TTDRecorder& _actionEntryPopper) -> JsErrorCode {
+        PERFORM_JSRT_TTD_RECORD_ACTION_NOT_IMPLEMENTED(scriptContext);
```
Comment (jianchun) [inline lib/Jsrt/Jsrt.cpp]: Do we need to implement this for TTD? @mrkmarron

**Comment by kfarnung** [comment]
Closing this one and retargeting to release/2.0 in #2594.

File: lib/Jsrt/Jsrt.cpp (hunk 5 above)
Comment (kfarnung) [inline lib/Jsrt/Jsrt.cpp]: I spoke with @mrkmarron and this was his recommendation.  If the user is running TTD and runs into this we'll fail fast and deal with it at that time.

**Pull request closed without merging**